src/bookshelf/src/bksw_sec.erl

```diff
@@ -21,123 +22,405 @@
+% very clear where the throws from this function and only this function are handled.>>
+%
+% WHAT GETS THROWN: well, a tuple... but it results in a 403 status code being returned.
+% WHY:              see explanation below.
+% WHERE HANDLED:    at the bottom of the function where the throws are occuring, within this module.
+%
+% Letting things crash upon errors would be nice here (i tried it), but it results in status code 500,
+% which causes failing tests.  It seems that in general, errors in processing here, e.g. no host or date header,
+% should return status code 403 vs. 500... or at least that's what various pre-existing tests want, anyway. The
+% throw is really just a way to bail on further processing and jump to a location in the code which returns a status 403.
+% Near as I can tell from the docs, that is what the purpose of throws are in Erlang.  Although
+% they are frequently associated with error handling, docs say they are for 'nonlocal returns'
+% or 'control flow' (or at least can be used that way).
+%
+% From: https://docs.oracle.com/cd/E88353_01/html/E37845/erlang-3erl.html
+%   throw(Any) -> no_return()
+%    A non-local return from a function. If evaluated within a catch,
+%    catch returns value Any. Example:
+%
+%    > catch throw({hello, there}).
+%    {hello,there}
+%
+% From: http://www.kuaidingyue.com/ebook/erl/errors-and-exceptions.html
+%   In comparison with exits and errors, they don't really carry any 'crash that process!' intent behind them,
+%   but rather control flow.
+%
+% Anyway, we could accomplish the same thing with a complicated nest or staircase of case statements, but to
+% me a simple throw is better. The catch which returns status code 403 is at the bottom of the auth
+% function.
+%
+% Having said all of that, if throws are undesirable, I'm totally open to eliminating them if we can figure out
+% how to do it.
+%
+% Perhaps one option is to rewrite the spec and say that 500s are ok.  This changes
+% specs and forces rewriting of tests, but avoids throws.
+%
+% Perhaps another option is to set variables to 'sentinel' values on error conditions,
+% which would allow the flow to continue, and hopefully the failures could be caught
+% later on in processing. An example might be setting Host (in auth/10 directly below
+% these comments) to "" if get_req_header is undefined, and letting things continue.
+% This might make things harder to debug, vs. an early bail with a throw (or who knows,
+% maybe early bails with throws are harder to debug?), but it avoids
+% throws if that is desirable.
+%
+% Or heck, maybe structuring case statements to eliminate throws wouldn't be as bad
+% as I am assuming. Or maybe I'm missing something.
+%
+% Anyway, I'm flexible.  For now, assuming throws until we get things sorted out.
+
+auth(RequestId, Req0, #context{reqid = ReqId} = Context, Credential, XAmzDate, SignedHeaderKeysString, IncomingSignature, XAmzExpiresString, Headers0, VerificationType) ->
```

Comment:
do we have tests for this function?

Reply:
No.  It wouldn't be easy to test in its current form, but I could investigate writing tests for it.